core/server/data/migrations/versions/2.27/3-add-subdirectory-to-relative-canonical-urls.js

```diff
@@ -0,0 +1,94 @@
+const models = require('../../../../models');
+const common = require('../../../../lib/common');
+const config = require('../../../../config');
+const {URL} = require('url');
+
+module.exports.config = {
+    transaction: true
+};
+
+// We've been incorrectly saving canonical_url fields without a subdirectory.
+// We need this column to be consistent with all other relative URLs, so...
+
+// if we have a configured url with a subdirectory
+// find all posts that have a canonical_url starting with / but not //
+// prefix the subdirectory to the canonical_url and save the post
+module.exports.up = (options) => {
+    const url = new URL(config.get('url'));
+
+    const localOptions = Object.assign({
+        context: {internal: true}
+    }, options);
+
+    if (url.pathname === '/') {
+        common.logging.info('Skipping posts.canonical_url subdirectory fix: no subdirectory configured');
+        return Promise.resolve();
+    }
+
+    // perform a specific query for the type of canonical URLs we're looking for
+    // so we're not fetching and manually looping over a ton of post models
+    return models.Posts
+        .forge()
+        .query((qb) => {
+            qb.where('canonical_url', 'like', '/%');
+            qb.whereNot('canonical_url', 'like', '//%');
+        })
+        .fetch(localOptions)
+        .then((posts) => {
+            if (posts) {
+                const queries = [];
+
+                posts.forEach((post) => {
+                    const canonicalUrl = post.get('canonical_url').replace('/', url.pathname);
+                    post.set('canonical_url', canonicalUrl);
+                    queries.push(post.save(null, localOptions));
+                });
+
+                return Promise.all(queries).then(() => {
+                    common.logging.info(`Added subdirectory prefix to canonical_url in ${posts.length} posts`);
+                });
+            }
+
+            common.logging.info('Skipping posts.canonical_url subdirectory fix: no canonical_urls to fix');
+        });
+};
+
+// if we have a configured url with a subdirectory
+// find all posts with a canonical_url starting with the subdirectory
+// remove it and save the post
+module.exports.down = (options) => {
+    const url = new URL(config.get('url'));
+
+    const localOptions = Object.assign({
+        context: {internal: true}
+    }, options);
+
+    if (url.pathname === '/') {
+        common.logging.info('Skipping posts.canonical_url subdirectory fix: no subdirectory configured');
+        return Promise.resolve();
+    }
+
+    return models.Posts
+        .forge()
+        .query((qb) => {
+            qb.where('canonical_url', 'LIKE', `${url.pathname}%`);
+        })
+        .fetch()
+        .then((posts) => {
+            if (posts) {
+                const queries = [];
+
+                posts.forEach((post) => {
+                    const canonicalUrl = post.get('canonical_url').replace(url.pathname, '/');
```

Comment:
Looks like there could be a bug here 🙈 `url.pathname` may or may not have a trailing slash depending what is put in the config file. Will test and update if necessary

Reply:
Resolved via normalisation here https://github.com/TryGhost/Ghost/pull/10989/files#diff-798de78c47b727a0903dec66e0802ce0R66-R69